### PR TITLE
[NCL-8386] Fix build metrics tooltip and refactor the component

### PR DESCRIPTION
**Pull request by DnsZhou**

### CODE REVIEW

**Review by patrikk0123 — APPROVED** [review]

**Comment by DnsZhou** [comment]
Hi @patrikk0123 , the code has been updated because the last version doesn't work well with the filter function(display every n builds). Please have a look at this as well. Thanks.

File: src/utils/utils.ts [hunk 1]
```diff
@@ -153,17 +153,16 @@ export const calculateDurationDiff = (diffMs: number | string): string | null =>
...
   const minutes = Math.floor((diffSec % 3600) / 60);
   const seconds = Math.floor(diffSec % 60);
-  const milliSeconds = Math.floor(diffMsValue % 1000000);
+  const milliSeconds = Math.floor(diffMsValue % 1000);
 
-  const daysString = days !== 0 ? days + 'd' : '';
-  const hoursString = hours !== 0 ? hours + 'h' : '';
-  const minutesString = days === 0 && minutes !== 0 ? minutes + 'm' : '';
-  const secondsString = days === 0 && hours === 0 && minutes === 0 && seconds !== 0 ? seconds + 's' : '';
-  const milliSecondsString = days === 0 && hours === 0 && minutes === 0 && seconds === 0 ? milliSeconds + 'ms' : '';
```
Comment (patrikk0123) [inline src/utils/utils.ts:162]: you removed milliseconds completely

File: src/utils/utils.ts [hunk 2]
```diff
@@ -283,3 +282,36 @@ export const getNumberGenerator = function* () {
...
+  let formattedSeconds: string;
+
+  if (minutes > 0 || hours > 0) {
+    formattedSeconds = `${Math.floor(secondsRemaining)}s`;
+  } else {
+    formattedSeconds = `${secondsRemaining.toFixed(2)}s`;
+  }
+
+  if (hours > 0) {
+    return `${hours}h ${minutesRemaining}m ${formattedSeconds} (${msTime}ms)`;
```
Comment (patrikk0123) [inline src/utils/utils.ts:309]: all `formattedSeconds` miss 's' unit

File: src/utils/utils.ts (hunk 1 above)
Comment (DnsZhou) [reply src/utils/utils.ts:162]: fixed it

File: src/utils/utils.ts (hunk 2 above)
Comment (DnsZhou) [reply src/utils/utils.ts]: fixed it

**Review by patrikk0123 — APPROVED** [review]

File: src/utils/utils.ts [hunk 3]
```diff
@@ -153,17 +153,17 @@ export const calculateDurationDiff = (diffMs: number | string): string | null =>
...
-  const milliSecondsString = days === 0 && hours === 0 && minutes === 0 && seconds === 0 ? milliSeconds + 'ms' : '';
+  const parts = [];
 
-  return `${daysString}${daysString && hoursString ? ' ' : ''}${hoursString}${
-    hoursString && minutesString ? ' ' : ''
-  }${minutesString}${secondsString}${milliSecondsString}`;
+  if (days) parts.push(`${days}d`);
+  if (hours) parts.push(`${hours}h`);
+  if (minutes) parts.push(`${minutes}m`);
+  if (seconds) parts.push(`${seconds}s`);
```
[the remark this answers is not in the document]
Comment (DnsZhou) [reply src/utils/utils.ts]: updated to show it conditionally

**Pull request merged**